src/Emitter.ts

```diff
@@ -0,0 +1,38 @@
+import { EventEmitter } from 'events';
+
+/**
+ * Mapping of event names to event handlers.
+ */
+type EventMap = {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [name: string]: (...args: any[]) => void;
+}
+
+export interface Emitter<Events extends EventMap> {
+  on<Name extends keyof Events>(event: Name, listener: Events[Name]): void;
+  emit<Name extends keyof Events>(event: Name, ...args: Parameters<Events[Name]>): void;
```

Comment:
that is some gnarly typescript 😵 I think I follow though (nice unit tests)

Reply:
Oh trust me, TypeScript can get way crazier than this (see https://gal.hagever.com/posts/typing-the-technical-interview-in-typescript/) 😂 

But this one is pretty "interesting" with two levels of generics with constraints.